### PR TITLE
supporting protobuf v22 and later(with abseil-cpp/C++17)

### PR DESCRIPTION
**Pull request by Kumataro**
fix https://github.com/opencv/opencv/issues/24369
related https://github.com/opencv/opencv/issues/23791

1. This patch supports external protobuf v22 and later, it required abseil-cpp and c++17.
    Even if the built-in protobuf is upgraded to v22 or later, 
    the dependency on abseil-cpp and the requirement for C++17 will continue.
2. Some test for caffe required patched protobuf, so this patch disable them.

This patch is tested by following libraries.
-  Protobuf:                    /usr/local/lib/libprotobuf.so (4.24.4)
-  abseil-cpp:                YES (20230125)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by Kumataro** [comment]
I was relieved that the test passed completely.

And this patch requires special setting to install protobuf v24+.

How to install external protobuf v24 and abseil-cpp at Ubuntu 23.04.
----------------------------------------------------------------------------

```bash
git clone https://github.com/protocolbuffers/protobuf.git
cd protobuf
  git checkout v4.24.4
  git reset --hard
  git submodule update --init --recursive
cd ..

cmake -S protobuf/third_party/abseil-cpp/ -B build_absl -GNinja -DCMAKE_CXX_STANDARD=17
cmake --build build_absl
sudo cmake --install build_absl
sudo ldconfig

cmake -S protobuf/ -B build_protobuf -GNinja -Dprotobuf_BUILD_TESTS=OFF -Dprotobuf_BUILD_SHARED_LIBS=ON -DCMAKE_CXX_STANDARD=17
cmake --build build_protobuf
sudo cmake --install build_protobuf
sudo ldconfig

cmake -S opencv -B build4-main_cxx17 -DCMAKE_CXX_STANDARD=17 -GNinja -DBUILD_PROTOBUF=OFF -DPROTOBUF_UPDATE_FILES=ON
```

To test building opencv4 with options.
-------------------------------------------

```bash
cmake -S opencv -B build4-main_cxx17 -DCMAKE_CXX_STANDARD=17 -GNinja -DBUILD_PROTOBUF=OFF -DPROTOBUF_UPDATE_FILES=ON
cmake --build main_cxx17 
```

If no patch, there are following errors.
------------------------------------------

```bash
/usr/bin/ld: modules/dnn/CMakeFiles/opencv_dnn.dir/attr_value.pb.cc.o: in function `absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <34>(char const (&) [34])':
attr_value.pb.cc:(.text._ZN4absl12lts_2023012512log_internal10LogMessagelsILi34EEERS2_RAT__Kc[_ZN4absl12lts_2023012512log_internal10LogMessagelsILi34EEERS2_RAT__Kc]+0x24): undefined reference to `absl::lts_20230125::log_internal::LogMessage::CopyToEncodedBuffer(std::basic_string_view<char, std::char_traits<char> >, absl::lts_20230125::log_internal::LogMessage::StringType)'
/usr/bin/ld: modules/dnn/CMakeFiles/opencv_dnn.dir/attr_value.pb.cc.o: in function `google::protobuf::MapKey::type() const [clone .part.0]':
...
```

File: modules/dnn/test/test_layers.cpp [hunk 1]
```diff
@@ -762,7 +762,11 @@ TEST_F(Layer_RNN_Test, get_set_test)
     EXPECT_EQ(shape(outputs[1]), shape(nT, nS, nH));
 }
 
+#ifndef OPENCV_DNN_EXTERNAL_PROTOBUF
 TEST_P(Test_Caffe_layers, Accum)
+#else
+TEST_P(Test_Caffe_layers, DISABLED_Accum)  // requires patched protobuf (available in OpenCV source tree only)
+#endif
```
Comment (dkurt) [inline modules/dnn/test/test_layers.cpp:769]: Skip exception is more preferable:
```cpp
TEST_P(Test_Caffe_layers, Accum)
{
#ifdef
  throw SkipTestException("Requires patched protobuf");
#else
    if (backend == DNN_BACKEND_OPENCV && target != DNN_TARGET_CPU)
        applyTestTag(CV_TEST_TAG_DNN_SKIP_OPENCL, CV_TEST_TAG_DNN_SKIP_OPENCL_FP16);

    testLayerUsingCaffeModels("accum", false, false, 0.0, 0.0, 2);
    testLayerUsingCaffeModels("accum_ref", false, false, 0.0, 0.0, 2);
#endif
}

File: modules/dnn/src/caffe/caffe_io.cpp [hunk 2]
```diff
@@ -1130,7 +1130,17 @@ bool ReadProtoFromTextFile(const char* filename, Message* proto) {
...
     parser.SetRecursionLimit(1000);
 #endif
-    return parser.Parse(&input, proto);
+    const bool ret = parser.Parse(&input, proto);
+
+#ifdef OPENCV_DNN_EXTERNAL_PROTOBUF
+    if(!ret)
+    {
+        LOG(ERROR) << "Some data requires patched protobuf (available in OpenCV source tree only).";
+        CV_Error_(Error::StsError,("Some data requires patched protobuf (available in OpenCV source tree only)."));
```
Comment (dkurt) [inline modules/dnn/src/caffe/caffe_io.cpp:1139]: Return value might be false even with patched protobuf. So I recommend to keep it as is.

File: cmake/OpenCVFindProtobuf.cmake [hunk 3]
```diff
@@ -89,3 +123,8 @@ if(HAVE_PROTOBUF)
     BUILD_PROTOBUF THEN "build (${Protobuf_VERSION})"
     ELSE "${__location} (${Protobuf_VERSION})")
 endif()
+
+if(HAVE_ABSL_STRINGS AND HAVE_ABSL_LOG)
+  list(APPEND CUSTOM_STATUS absl)
+  list(APPEND CUSTOM_STATUS_absl "      abseil-cpp:" "YES (${ABSL_STRINGS_VERSION})" )
```
Comment (dkurt) [inline cmake/OpenCVFindProtobuf.cmake:129]: `abseil-cpp` is a direct dependency of Protobuf, not OpenCV. Not sure that it has to be reported.

File: modules/dnn/test/test_layers.cpp (hunk 1 above)
Comment (Kumataro) [reply modules/dnn/test/test_layers.cpp:769]: Thank you for review!
I implemented new patch for some layers to disable, which is same as the already implemented tests for `interp`.
Is my recognition that the test for Interp should be modified in the same way correct?

https://github.com/opencv/opencv/blob/590f150d5e032165e27d81294c9b7ac710b77f11/modules/dnn/test/test_layers.cpp#L1654-L1658

Comment (dkurt) [reply modules/dnn/test/test_layers.cpp:769]: Yes, please. It is just a convenient way.

File: cmake/OpenCVFindProtobuf.cmake (hunk 3 above)
Comment (Kumataro) [reply cmake/OpenCVFindProtobuf.cmake:129]: When using protobuf v22 or later, 
the dnn module directly references `abseil-cpp` library ( via *.pb.cc converted *.pb with `protoc` ).

Unfortunally, there is also a possibility that OpenCV Users may use a combination of `protobuf` and `abseil-cpp` versions that are not expected.

I think it is better to leave abseil-cpp version information as a cmake log.
It is helpful to investifate compile/runtime errors

```bash
kmtr@kmtr-VMware-Virtual-Platform:~/work/build4-main_cxx17/modules/dnn/CMakeFiles/opencv_dnn.dir$ nm attr_value.pb.cc.o | c++filt | grep "absl" | head -10
                 U absl::lts_20230125::log_internal::LogMessage::OstreamView::stream()
                 U absl::lts_20230125::log_internal::LogMessage::OstreamView::OstreamView(absl::lts_20230125::log_internal::LogMessage::LogMessageData&)
                 U absl::lts_20230125::log_internal::LogMessage::OstreamView::~OstreamView()
                 U absl::lts_20230125::log_internal::LogMessage::CopyToEncodedBuffer(std::basic_string_view<char, std::char_traits<char> >, absl::lts_20230125::log_internal::LogMessage::StringType)
0000000000000000 W absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <14>(char const (&) [14])
0000000000000000 W absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <22>(char const (&) [22])
0000000000000000 W absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <23>(char const (&) [23])
0000000000000000 W absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <2>(char const (&) [2])
0000000000000000 W absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <34>(char const (&) [34])
0000000000000000 t absl::lts_20230125::log_internal::LogMessage& absl::lts_20230125::log_internal::LogMessage::operator<< <48>(char const (&) [48]) [clone .isra.0]
```

File: modules/dnn/src/caffe/caffe_io.cpp (hunk 2 above)
Comment (Kumataro) [reply modules/dnn/src/caffe/caffe_io.cpp:1139]: Thank you for the review! I would like to revert it.

Currently, there is no way to distinguish between (1) and (2).

- (1) data that can be processed by only internal protobuf.
- (2) data that cannot be processed by both internal and external protobuf.

-----
I think this patch we can know which protobuf the issue reporter is used.

If an external unpatched protobuf, we can suggest trying the internal protobuf.
If a patched internal protobuf, we can focus our investigation on data content, etc.

However this is only workaround, so I agree with you.

File: cmake/OpenCVFindProtobuf.cmake (hunk 3 above)
Comment (Kumataro) [reply cmake/OpenCVFindProtobuf.cmake:129]: I thought about it additionally.

When a user builds a protobuf, it is fully expected that the appropriate abseil-cpp version will be applied in the git submodule.

Therefore, I would like to delete the version log. thank you very much!

e.g.  https://github.com/protocolbuffers/protobuf/tree/main/third_party

> abseil-cpp @ fb3621f  Update to latest absl LTS patch 20230802.1 (#14145)

File: modules/dnn/src/caffe/caffe_io.cpp (hunk 2 above)
Comment (Kumataro) [reply modules/dnn/src/caffe/caffe_io.cpp]: I reverted it, thank you for your comment !

File: modules/dnn/CMakeLists.txt [hunk 4]
```diff
@@ -240,6 +249,10 @@ ocv_create_module(${libs} ${dnn_runtime_libs})
 ocv_add_samples()
 ocv_add_accuracy_tests(${dnn_runtime_libs})
 
+if(NOT BUILD_PROTOBUF)
+  ocv_target_compile_definitions(opencv_test_dnn PRIVATE "OPENCV_DNN_EXTERNAL_PROTOBUF=1")
+endif()
```
Comment (dkurt) [inline modules/dnn/CMakeLists.txt:254]: Does it mean that `DISABLED_Interp` was always disabled? There was `#ifndef OPENCV_DNN_EXTERNAL_PROTOBUF` but no definition?

Comment (Kumataro) [reply modules/dnn/CMakeLists.txt:254]: `OPENCV_DNN_EXTERNAL_PROTOBUF` definition seems not be available for `opencv_test_dnn` in currently implementation.

Old test result with external protobuf (not including this change) is following. If that definition is available, test for `Interp` is disabled. But the test  is executed.

So I think an additional fix is needed.

Code

https://github.com/opencv/opencv/blob/590f150d5e032165e27d81294c9b7ac710b77f11/modules/dnn/test/test_layers.cpp#L1654-L1658

Result

```plaintext
[       OK ] Test_Caffe_layers.PriorBox_squares/0 (0 ms)
[ RUN      ] Test_Caffe_layers.Interp/0, where GetParam() = OCV/CPU
E0000 00:00:1696641648.291055   46844 text_format.cc:433] Error parsing text-format opencv_caffe.NetParameter: 14:16: Message type "opencv_caffe.LayerParameter" has no field named "interp_param".
unknown file: Failure
C++ exception with description "OpenCV(4.8.0-dev) /home/kmtr/work/opencv/modules/dnn/src/caffe/caffe_io.cpp:1162: error: (-2:Unspecified error) FAILED: ReadProtoFromTextFile(param_file, param). Failed to parse NetParameter file: /home/kmtr/work/opencv_extra/testdata/dnn/layers/layer_interp.prototxt in function 'ReadNetParamsFromTextFileOrDie'
" thrown in the test body.
[  FAILED  ] Test_Caffe_layers.Interp/0, where GetParam() = OCV/CPU (1 ms)
[----------] 31 tests from Test_Caffe_layers (93 ms total)

[----------] Global test environment tear-down
[==========] 31 tests from 1 test case ran. (93 ms total)
[  PASSED  ] 26 tests.
```

Comment (dkurt) [reply modules/dnn/CMakeLists.txt:254]: Good catch

File: modules/dnn/CMakeLists.txt [hunk 5]
```diff
@@ -249,6 +249,10 @@ ocv_create_module(${libs} ${dnn_runtime_libs})
 ocv_add_samples()
 ocv_add_accuracy_tests(${dnn_runtime_libs})
 
+if(NOT BUILD_PROTOBUF)
+  ocv_target_compile_definitions(opencv_test_dnn PRIVATE "OPENCV_DNN_EXTERNAL_PROTOBUF=1")
```
Comment (shr-project) [inline modules/dnn/CMakeLists.txt:253]: Does this need "if(TARGET opencv_test_dnn)" as well? I'm seeing:

```
CMake Error at cmake/OpenCVUtils.cmake:1517 (message):
  ocv_target_compile_definitions: invalid target: 'opencv_test_dnn'
Call Stack (most recent call first):
  modules/dnn/CMakeLists.txt:258 (ocv_target_compile_definitions)
```

Comment (Kumataro) [reply modules/dnn/CMakeLists.txt]: Hello, thank you for your comment, I agree with you.
When `ENABLE_TESTS=OFF`, same problem is occured.
I push additional patch to fix it.

**Comment by shr-project** [comment]
Thanks, tested in OpenEmbedded environment where I've first seen this issue and with these changes it builds fine again. Thank you!

File: cmake/OpenCVFindProtobuf.cmake [hunk 6]
```diff
@@ -67,6 +67,40 @@ else()
...
+        if(REQUEST_ABSL)
+            string(COMPARE GREATER "${CMAKE_CXX_STANDARD}" "16" USED_AFTER_CXX17)  # >=17
+
+            if(NOT USED_AFTER_CXX17)
+                message("CMAKE_CXX_STANDARD : ${CMAKE_CXX_STANDARD}")
+                message("protobuf           : ${Protobuf_VERSION}")
+                message(FATAL_ERROR "protobuf(v22 and later) and abseil-cpp request CMAKE_CXX_STANDARD=17 and later.")
+            endif()
+
+            ocv_check_modules(ABSL_STRINGS absl_strings)
```
Comment (mshabunin) [inline cmake/OpenCVFindProtobuf.cmake:88]: Does abseil has cmake package? Can we use `find_package` instead of `pkg_config` and `find_library`?

**Comment by mshabunin** [comment]
Some time ago I've been able to build OpenCV with latest protobuf: https://github.com/opencv/opencv/issues/23791#issuecomment-1597579393 And I did not need to explicitly link with abseil. Did something change over time or was my configuration different than yours in some way?

**Comment by Kumataro** [comment]
> Maybe you don't need to build abseil separately then it will be built as part of protobuf and will be linked automatically?

Thank you for your comment.  

I found https://protobuf.dev/support/migration/#abseil

There are two ways protobuf references abseil-cpp.
1. `protobuf_ABSL_PROVIDER=module` (abseil-cpp will be included in protobuf)
2. `protobuf_ABSL_PROVIDER=package` (protobuf refers external abseil-cpp)

If `abseil-cpp` is installed manually in the system before running `cmake` for `protobuf`, option 2, that is using external `abseil-cpp` package, is selected. 

Sorry, I'll need some time to work.
My current idea is ....

1. Search for `abseil-cpp` with `find_package()`.
2. If found, add `abseil string` and `abseil log` to the link targets of the dnn module with `protobuf`
3. (If not ) Only `protobuf` will be linked for the dnn module.

**Comment by mshabunin** [comment]
I tried to build protobuf@24.x and OpenCV. As in previous result only minor patch was needed (replace `Protobuf` prefix with `protobuf` in two places, see link in my previous comment).  Apparently when using `Protobuf` name, cmake's own search method will be used which does not set imported targets property correctly, that is why it is necessary to search for `protobuf` package.

protobuf cmake package already searches for abseil (see `protobuf-config.cmake`):
```.cmake
if(NOT TARGET absl::strings)
  find_package(absl CONFIG)
endif()
```
and imported target already provides required interface link libraries (see `protobuf-targets.cmake`):
```.cmake
# Create imported target protobuf::libprotobuf
add_library(protobuf::libprotobuf SHARED IMPORTED)

set_target_properties(protobuf::libprotobuf PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "PROTOBUF_USE_DLLS"
  INTERFACE_COMPILE_FEATURES "cxx_std_14"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "absl::absl_check;absl::absl_log;absl::algorithm;absl::base;absl::bind_front;absl::bits;absl::btree;absl::cleanup;absl::cord;absl::core_headers;absl::debugging;absl::die_if_null;absl::dynamic_annotations;absl::flags;absl::flat_hash_map;absl::flat_hash_set;absl::function_ref;absl::hash;absl::layout;absl::log_initialize;absl::log_severity;absl::memory;absl::node_hash_map;absl::node_hash_set;absl::optional;absl::span;absl::status;absl::statusor;absl::strings;absl::synchronization;absl::time;absl::type_traits;absl::utility;absl::variant"
)
```

My script:
```.sh
PD=/work/protobuf/.build

build_pb() {
    rm -rf ${PD}
    mkdir -p ${PD} && pushd ${PD}
    cmake -GNinja \
        -DCMAKE_BUILD_TYPE=Release \
        -Dprotobuf_BUILD_TESTS=OFF \
        -DCMAKE_POSITION_INDEPENDENT_CODE=ON \ # need to be compatible with dynamic linkage
        -DCMAKE_INSTALL_PREFIX=install \
        -Dprotobuf_BUILD_SHARED_LIBS=ON \ # static works as well
        ..
    ninja install
    popd
}

build_opencv() {
    export LD_LIBRARY_PATH=${PD}/install/lib:${LD_LIBRARY_PATH} # need to set this for protoc to work
    mkdir -p $D
    pushd $D && rm -rf *
    cmake \
        -DCMAKE_BUILD_TYPE=Release \
        -DCMAKE_PREFIX_PATH=${PD}/install \
        -DWITH_PROTOBUF=ON \
        -DBUILD_PROTOBUF=OFF \
        -DPROTOBUF_UPDATE_FILES=ON \
        -Dprotobuf_MODULE_COMPATIBLE=ON \ # we still need this for command PROTOBUF_GENERATE_CPP
        -DCMAKE_CXX_STANDARD=17 \
        -DCMAKE_BUILD_TYPE=Release \
        -DWITH_OPENCL=OFF \
        ../opencv
    ninja
    popd
}

build_pb
build_opencv
```

In previous PR I promised to integrate this patch to OpenCV, but completely forgot about it :facepalm:

**Comment by Kumataro** [comment]
I'm sorry to not finished this fix.
It seems that my trial to remove direct-referencing `absl-cpp` libraries doesn't work well.
I think my operation maybe has some mistakes. I'll re-think about it calmly tomorrow(today).

I tried with Ubuntu23.04.
- (1) `libs` in modules/dnn/CMakefile is "ocv.3rdparty.flatbuffers;protobuf::libprotobuf". It is OK.
- (2) when linking dnn module, the referenced external library is only `libprotobuf.so`, not absl-cpp. So there are some undefined symbols.
- (3) `pkg-config protobuf --libs` outputs with absl libraries.

(1)
```
-- opencv_dnn: filter out cuda4dnn source code
-- opencv_dnn: ocv_create_module( ocv.3rdparty.flatbuffers protobuf::libprotobuf )
-- Excluding from source files list: <BUILD>/modules/dnn/layers/layers_common.avx.cpp
```

(2)
```
 modules/dnn/CMakeFiles/opencv_dnn.dir/opencl_kernels_dnn.cpp.o
 modules/dnn/CMakeFiles/opencv_dnn.dir/layers/cpu_kernels/fast_gemm_kernels.neon.cpp.o
  -Wl,-rpath,/home/kmtr/work/build4-main_cxx17/lib:/usr/local/lib:  
lib/libopencv_imgproc.so.4.8.0  -ldl  -lm  -lpthread  -lrt  
3rdparty/lib/libtegra_hal.a  
/usr/local/lib/libprotobuf.so  
lib/libopencv_core.so.4.8.0 && :
```

(3)
```
kmtr@ubuntu:~/work/build4-main_cxx17$ pkg-config protobuf --libs
-L/usr/local/lib -lprotobuf -labsl_log_internal_check_op -labsl_leak_check -labsl_die_if_null -labsl_log_internal_conditions -labsl_log_internal_message -labsl_examine_stack -labsl_log_internal_format -labsl_log_internal_proto -labsl_log_internal_nullguard -labsl_log_internal_log_sink_set -labsl_log_sink -labsl_log_entry -labsl_flags -labsl_flags_internal -labsl_flags_marshalling -labsl_flags_reflection -labsl_flags_private_handle_accessor -labsl_flags_commandlineflag -labsl_flags_commandlineflag_internal -labsl_flags_config -labsl_flags_program_name -labsl_log_initialize -labsl_log_globals -labsl_log_internal_globals -labsl_hash -labsl_city -labsl_low_level_hash -labsl_raw_hash_set -labsl_hashtablez_sampler -labsl_statusor -labsl_status -labsl_cord -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_exponential_biased -labsl_cordz_handle -labsl_crc_cord_state -labsl_crc32c -labsl_crc_internal -labsl_crc_cpu_detect -labsl_bad_optional_access -labsl_str_format_internal -labsl_strerror -labsl_synchronization -labsl_graphcycles_internal -labsl_stacktrace -labsl_symbolize -labsl_debugging_internal -labsl_demangle_internal -labsl_malloc_internal -labsl_time -labsl_civil_time -labsl_time_zone -labsl_bad_variant_access -lutf8_validity -lutf8_range -labsl_strings -labsl_strings_internal -lrt -labsl_base -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_raw_logging_internal -labsl_log_severity
```

**Comment by Kumataro** [comment]
Hello, I made the workaround clearly. ( I'll fix it after my job.. )

I tried to debug following command to check `protobuf-targets.cmake` is available.

```
ocv_glob_module_sources(${sources_options} SOURCES ${fw_srcs} ${webnn_srcs})

message("${libs}")
include(CMakePrintHelpers)
CMAKE_PRINT_PROPERTIES(TARGETS "protobuf::libprotobuf" PROPERTIES INTERFACE_COMPILE_DEFINITIONS INTERFACE_COMPILE_FEA
TURES INTERFACE_INCLUDE_DIRECTORIES INTERFACE_LINK_LIBRARIES)

ocv_create_module(${libs} ${dnn_runtime_libs})
```

`find_package(Protobuf QUIET CONFIG)` is required.

If there is "CONFIG" in `find_package()`, we can get nessary properties.

```
 Properties for TARGET protobuf::libprotobuf:
   protobuf::libprotobuf.INTERFACE_COMPILE_DEFINITIONS = "PROTOBUF_USE_DLLS"
   protobuf::libprotobuf.INTERFACE_COMPILE_FEA = <NOTFOUND>
   protobuf::libprotobuf.TURES = <NOTFOUND>
   protobuf::libprotobuf.INTERFACE_INCLUDE_DIRECTORIES = "/usr/local/include"
   protobuf::libprotobuf.INTERFACE_LINK_LIBRARIES = "absl::absl_check;absl::absl_log;absl::algorithm;absl::base;absl::bind_front;absl::bits;absl::btree;absl::cleanup;absl::cord;absl::core_headers;absl::debugging;absl::die_if_null;absl::dynamic_annotations;absl::flags;absl::flat_hash_map;absl::flat_hash_set;absl::function_ref;absl::hash;absl::layout;absl::log_initialize;absl::log_severity;absl::memory;absl::node_hash_map;absl::node_hash_set;absl::optional;absl::span;absl::status;absl::statusor;absl::strings;absl::synchronization;absl::time;absl::type_traits;absl::utility;absl::variant"
```

if NOT...

```
 Properties for TARGET protobuf::libprotobuf:
   protobuf::libprotobuf.INTERFACE_COMPILE_DEFINITIONS = <NOTFOUND>
   protobuf::libprotobuf.INTERFACE_COMPILE_FEA = <NOTFOUND>
   protobuf::libprotobuf.TURES = <NOTFOUND>
   protobuf::libprotobuf.INTERFACE_INCLUDE_DIRECTORIES = "/usr/local/include"
   protobuf::libprotobuf.INTERFACE_LINK_LIBRARIES = "Threads::Threads"
```

Now my test code doesn't reference `absl-cpp` libraries directly. It works well.

**Comment by mshabunin** [comment]
> If there is "CONFIG" in find_package(), we can get nessary properties.

As I understood this is related to cmake's built-in protobuf search mechanism. It seem to be using pkg-config and can not correctly add abseil dependencies. Adding `CONFIG` or `NO_MODULE` or specifying name `protobuf` instead of `Protobuf` disables built-in search mechanism and uses actual cmake package provided by protobuf.

I'm working on an integration testing harness for protobuf, I'll try to provide some solution today or tomorrow.

**Comment by Kumataro** [comment]
I believe it is better to wait your comments/suggestions. Thank you very much !

This problem ,which is related with linking `abseil-cpp`, is discussed at https://github.com/protocolbuffers/protobuf/issues/12292 (Protobuf Github)

However, it is not resolved clearly, it seems that there are many problems.

**Comment by Kumataro** [comment]
I intended to maintain the old branch and commit the code without dependence on ABSL to the new branch.

However, I made a mistake updating the same branch. The modifications are as expected. I'm sorry...

File: cmake/OpenCVFindProtobuf.cmake [hunk 7]
```diff
@@ -67,22 +74,51 @@ else()
...
+    if("${Protobuf_VERSION}" MATCHES [[[0-9]+.([0-9]+).[0-9]+]])
+        string(COMPARE GREATER "${CMAKE_MATCH_1}" "21" REQUEST_AFTER_CXX17)  # >=22
+
+        if(REQUEST_AFTER_CXX17)
+            string(COMPARE GREATER "${CMAKE_CXX_STANDARD}" "16" USED_AFTER_CXX17)  # >=17
+
+            if(NOT USED_AFTER_CXX17)
+                message("CMAKE_CXX_STANDARD : ${CMAKE_CXX_STANDARD}")
+                message("protobuf           : ${Protobuf_VERSION}")
+                message(FATAL_ERROR "protobuf(v22 and later) and abseil-cpp request CMAKE_CXX_STANDARD=17 and later.")
```
Comment (mshabunin) [inline cmake/OpenCVFindProtobuf.cmake:92]: Usually OpenCV does not throw errors if an optional dependency can not be found or used. @asmorkalov , @opencv-alalek , what do you think? Should we show a warning and disable protobuf?

**Review by mshabunin — COMMENTED** [review]
I've tested this PR using several different protobuf versions (https://github.com/mshabunin/protobuf-check), builds are fine, test had not run.

Overall looks good to me :+1: , except one question.

File: cmake/OpenCVFindProtobuf.cmake (hunk 7 above)
Comment (Kumataro) [reply cmake/OpenCVFindProtobuf.cmake:92]: Thank you for your review !
I would like to wait for comments for the direction of correction.

And unfortunately, there are some `message(FATAL_ERROR)` in this file like this block. If error is occured, they will stop configuration.

https://github.com/opencv/opencv/blob/5199850039ad23f1f0e6cccea5061a9fea5efca6/modules/dnn/CMakeLists.txt#L133-L135

https://github.com/opencv/opencv/blob/5199850039ad23f1f0e6cccea5061a9fea5efca6/modules/dnn/CMakeLists.txt#L166-L168

https://github.com/opencv/opencv/blob/5199850039ad23f1f0e6cccea5061a9fea5efca6/modules/dnn/CMakeLists.txt#L228-L230

File: cmake/OpenCVFindProtobuf.cmake [hunk 8]
```diff
@@ -67,22 +74,51 @@ else()
...
 endif()
 
+# See https://github.com/opencv/opencv/issues/24369
+# In Protocol Buffers v22.0 and later drops C++11 support and depends abseil-cpp.
+#   Details: https://protobuf.dev/news/2022-08-03/
+# And if std::text_view is in abseil-cpp requests C++17 and later.
+
+if(HAVE_PROTOBUF)
+    if("${Protobuf_VERSION}" MATCHES [[[0-9]+.([0-9]+).[0-9]+]])
+        string(COMPARE GREATER "${CMAKE_MATCH_1}" "21" REQUEST_AFTER_CXX17)  # >=22
```
Comment (opencv-alalek) [inline cmake/OpenCVFindProtobuf.cmake:84]: What is the problem with `if'(NOT (Protobuf_VERSION VERSION_LESS 22))`?

https://cmake.org/cmake/help/latest/command/if.html#version-less

> Any non-integer version component or non-integer trailing part of a version component effectively truncates the string at that point.

File: cmake/OpenCVFindProtobuf.cmake [hunk 9]
```diff
@@ -67,22 +74,51 @@ else()
...
+
+        if(REQUEST_AFTER_CXX17)
+            string(COMPARE GREATER "${CMAKE_CXX_STANDARD}" "16" USED_AFTER_CXX17)  # >=17
+
+            if(NOT USED_AFTER_CXX17)
```
Comment (opencv-alalek) [inline cmake/OpenCVFindProtobuf.cmake:89]: Use `if(GREATER)` instead.

File: cmake/OpenCVFindProtobuf.cmake (hunk 7 above)
Comment (opencv-alalek) [reply cmake/OpenCVFindProtobuf.cmake:92]: It is not a default configuration.
Need to explicitly specify `BUILD_PROTOBUF=OFF` to go here.
I believe we could these errors here.

Comment (Kumataro) [reply cmake/OpenCVFindProtobuf.cmake]: Thank you for your review!
Yes, it is clear. This is about `protobuf`, so I set HAVE_PROTOBUF=OFF if old cxx standard is used. (And I forget C++98 ...)

File: cmake/OpenCVFindProtobuf.cmake (hunk 8 above)
Comment (Kumataro) [reply cmake/OpenCVFindProtobuf.cmake]: Yes, It is better than my script. Thank you very much !

File: cmake/OpenCVFindProtobuf.cmake (hunk 9 above)
Comment (Kumataro) [reply cmake/OpenCVFindProtobuf.cmake]: This comparing result is used with "NOT", so I use `LESS`.
And I add to check C++98. Thank you !

**Review by mshabunin — APPROVED** [review]

File: cmake/OpenCVFindProtobuf.cmake [hunk 10]
```diff
@@ -6,6 +6,7 @@ if(NOT WITH_PROTOBUF)
   return()
 endif()
 
+
```
Comment (Kumataro) [inline cmake/OpenCVFindProtobuf.cmake]: I'm sorry I have small mistake that is inserting unnecessary blank line.I had created pull request to fix it.

**Review by shr-project — APPROVED** [review]

**Comment by asmorkalov** [comment]
@dkurt @opencv-alalek May I merge the patch?

**Review by dkurt — APPROVED** [review]

**Review by opencv-alalek — APPROVED** [review]
LGTM :+1:

**Pull request merged**